eggtimer.js

```diff
@@ -1,326 +1,681 @@
-    const url = event.payload.pull_request.url;
-    const head_sha = event.payload.pull_request.head.sha;
-    const ref = event.payload.pull_request.head.ref;
-    console.log(url + " -> pull_request");
-    ensurePr(url, head_sha);
-    prs[url].ref = ref;
-    populateMergeable(url);
-    populateReviews(url);
-    mergeIfReady(url);
+WebhookHandler.on('pull_request', (ev) => {
+    const pr = ev.payload.pull_request;
+    console.log("pull_request event:", ev.payload.id, pr.number, pr.head.sha, pr.state);
+    processEvent();
 });
 
 // https://developer.github.com/v3/activity/events/types/#statusevent
-HANDLER.on('status', function(event) {
-    const sha = event.payload.sha;
-    const context = event.payload.context;
-    const state = event.payload.state;
-    var success = false;
-    switch (state) {
-        case 'success':
-            success = true;
-            break;
-        case 'pending':
-        case 'failure':
-        case 'error':
-            // success = false, still
-            break;
-        default:
-            console.error("Unknown check state '" + state + "'. success = false");
-            break;
+WebhookHandler.on('status', (ev) => {
+    const e = ev.payload;
+    console.log("status event:", e.id, e.sha, e.context, e.state);
+    if (merging(e.sha)) {
+        currentContext.signaled = true;
+        finishMerging(currentContext);
+    } else
+        processEvent();
+});
+
+// startup
+initContext();
+
+function merging(sha) {
+    return (currentContext !== null && currentContext.autoSha !== null && sha === currentContext.autoSha);
+}
+
+function initContext() {
+    assert(currentContext === null);
+    currentContext = {};
+    let autoSha = null;
+
+    let autoParams = commonParams();
+    autoParams.ref = "heads/" + Config.auto_branch;
+    let autoPromise = getReference(autoParams);
+
+    let tagsParams = commonParams();
+    let tagsPromise = getTags(tagsParams);
+
+    Promise.all([autoPromise, tagsPromise])
+        .then( (results) => {
+            autoSha = results[0];
+            let tags = results[1];
+            let prNum = null;
+
+            tags.find( (tag) => {
+                if (tag.object.sha === autoSha) {
+                    let matched = tag.ref.match(TagRegex);
+                    if (matched)
+                        prNum = matched[2];
+                }
+            });
+
+            if (prNum === null)
+                return Promise.reject(true);
+
+            let params = commonParams();
+            params.number = prNum;
+            return getPR(params);
+        })
+        .then((pr) => {
+            currentContext = createPRContext(pr);
+            currentContext.autoSha = autoSha;
+            console.log("Found PR" + pr.number + " in a merging state");
+            finishMerging(currentContext);
+            return;
+        })
+        .catch((err) => {
+            console.log("No PR is in the merging state: ", err);
```

Comment:
Though this is an exception, it is not an error, because I used 'console.log()' instead of 'console.error()'. Using exceptions helps to stop later processing exit immediately, I don't understand what is bad in using them.

Reply:
Using exceptions for normal/desirable/anticipated/regular conditions conflicts with the exception intended purpose/design. Such misapplication of standard development concepts may confuse developers, make code more complex than necessary, slow down the program, and lead to bugs. We should use exceptions for their intended purpose unless there is a very good (and clear or documented) reason to violate this natural rule of thumb.